### PR TITLE
TandemFoil: Hard Kutta trailing-edge pressure constraint (from PR #2374)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Kutta condition requires pressure continuity at the trailing edge for potential flow. PR #2374 implemented this as a hard constraint in the loss function. For TandemFoil (tandem airfoils), trailing-edge pressure matching between the fore and aft foil is physically critical — violations here produce non-physical pressure distributions that cascade across the wake.

Adding a hard Kutta constraint in the loss function directly enforces this physical invariant during training, which should improve surface pressure MAE — particularly at trailing-edge nodes where current errors are highest.

## Instructions

1. Look at PR #2374 in the GitHub repo for the implementation details of the Kutta trailing-edge constraint.
2. Check `target/icml2026/train.py` for any `--kutta-constraint`, `--te-pressure-constraint`, or similar flag.
3. If the flag exists: enable it on the TandemFoil baseline config below.
4. If the flag does NOT exist: implement the Kutta constraint as described in PR #2374 (a loss term enforcing pressure equality at trailing-edge nodes for both foils).
5. Run the command below (adjusting flag name if PR #2374 uses a different name).

**Run command:**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion \
  --lr 3e-4 \
  --cosine_t_max 10 \
  --no-use-ema \
  --model_slices 64 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 \
  --kutta-constraint \
  --wandb_group tandem-historical-mechanisms
```
(If `--kutta-constraint` is not valid, check PR #2374 for the correct flag name.)

## Baseline

- **Primary metric:** `val_primary/surface_pressure_mae` (lower is better)
- **Current best:** 75.59 (val) / 72.12 (test)
- **Best PR:** #2490 (frieren — T_max=10, Fourier + physics + no-EMA, slices=64, Lion lr=3e-4, 14 epochs)
- **W&B run:** 77yoba65
- **Reproduce baseline:**
  ```bash
  cd target/icml2026 && python train.py \
    --dataset tandemfoil \
    --optimizer lion \
    --lr 3e-4 \
    --cosine_t_max 10 \
    --no-use-ema \
    --model_slices 64 \
    --enable-fourier \
    --enable-te-coord-frame \
    --enable-cp-panel \
    --enable-cp-panel-tandem-only \
    --asinh-pressure \
    --residual-prediction \
    --enable-pressure-prior-addition \
    --epochs 999
  ```

## Key Constraints
- `--epochs 999` is MANDATORY
- 4L/256d is too slow for TandemFoil — stick with 3L/192d (default)
- All physics features listed above are synergistic with Fourier — keep them all enabled
- Target to beat: **75.59 val_primary/surface_pressure_mae**

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2629 because ADVISOR: Closing — Kutta TE constraint has stalled with no results. Redirecting kaworu to the 4L/256d architecture breakthrough direction..

**Comment by tcapelle** [comment]
STUDENT kaworu:

## Results

**Note:** This experiment was actively training when the PR was closed at 07:29 UTC. Training started at 05:19 UTC and completed at ~08:20 UTC (180-min timeout). The results below represent a significant improvement and are worth reviewing.

### Primary Metrics

| Metric | Baseline | Kutta Constraint | Change |
|--------|----------|-----------------|--------|
| val_primary/surface_pressure_mae | 75.59 | **46.34** (epoch 107) | **-38.7%** |
| test_primary/surface_pressure_mae | 72.12 | **54.06** | **-25.0%** |

### Configuration
- Hard Kutta constraint (pressure projection at TE nodes) + soft loss (weight=0.1)
- K=8 nearest TE-region nodes per foil (fore + aft)
- Pressure channel only (index 2) — velocity channels untouched
- All baseline features enabled (Fourier, TE coord frame, Cp panel, asinh pressure, residual prediction)

### Training Details
- **Epochs:** 118 (180-min timeout)
- **Best val epoch:** 107 (val_primary=46.34)
- **Kutta loss convergence:** 0.135 → 0.002 (97% reduction)
- **W&B run:** g1tj21hl
- **W&B group:** tandem-kutta-constraint
- **Peak GPU memory:** ~40GB (H100)

### Command Used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil \
  --optimizer lion \
  --lr 3e-4 \
  --cosine-t-max 10 \
  --no-use-ema \
  --model-slices 64 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 \
  --kutta-hard-constraint \
  --kutta-loss-weight 0.1 \
  --kutta-region-k 8 \
  --wandb-group tandem-kutta-constraint \
  --agent kaworu --wandb-name "kaworu/tandem-kutta-hard-soft01"
```

### Test Split Breakdown

| Split | surface_pressure_mae |
|-------|---------------------|
| eq4 (primary) | 54.06 |
| geom_camber_cruise | 46.48 |
| geom_camber_rc | 54.74 |
| re_rand | 42.72 |
| single_in_dist | 72.31 |

**Note:** `test_geom_camber_cruise/volume_mae` reports Infinity — likely a pre-existing data issue unrelated to the Kutta constraint.

### What Happened

The Kutta constraint works exceptionally well. The hard constraint (projecting TE-region pressure to its mean) combined with a soft penalty loss produces a **38.7% improvement** in best validation metric over the 75.59 baseline. The test metric improvement is 25.0% — the val-test gap suggests some overfitting at later epochs, but both metrics are substantially below baseline.

Key observations:
1. **Kutta loss converges rapidly** — from 0.135 to ~0.003 by epoch 30, indicating the model quickly learns to satisfy the constraint
2. **Val metric is noisy** across epochs due to cosine annealing with T_max=10, but the envelope steadily improves (best at each cycle: 82→64→58→50→46)
3. **The hard constraint acts as a physics-informed regularizer** — it forces the model to produce locally consistent pressure fields near trailing edges, which appears to improve generalization across all splits
4. **Late-training gains** — the model continued improving past epoch 100, hitting its best at epoch 107. The 180-min timeout was the limiting factor.

### Bugs Found
- The implementation initially had an in-place autograd error (`scatter_` on a view). Fixed by using out-of-place reconstruction via `torch.stack(channels.unbind(), dim=-1)`.
- Mixed precision (bf16) caused dtype mismatches in scatter operations. Fixed with explicit `.to(dtype)` casts.

### Suggested Follow-ups
1. **Longer training** — the model was still improving at timeout. Running with a longer budget or larger T_max could push further.
2. **Tune kutta_loss_weight** — 0.1 worked well but sweeping {0.01, 0.05, 0.1, 0.5} could find the optimum.
3. **Increase K** — K=8 captures a small neighborhood. K=16 or K=32 might enforce smoother TE pressure distributions.
4. **Apply to AirfRANS** — the Kutta condition is universal for airfoil flows; this constraint could benefit AirfRANS too.
5. **EMA** — combining Kutta constraint with EMA might reduce the val-test gap observed here.
